### PR TITLE
Create SubmissionDaoTests file

### PR DESCRIPTION
**Pull request by frozenfrank**
## Overview

Introduce tests to verify that our DAOs work properly. This gives us confidence that things work properly as changes are made.

## Details

- I was recently making sweeping changes to the Submission object, and I wanted to be able to verify that they were working properly. This didn't help me at that time, but I want to establish a precedent and a pattern for future developers to use to test DAOs.
- I need help getting the tests to run. It requires a database config, which I apparently do have set up properly yet.
    - There may be a better way setup a SQL server for testing.
- If this works, we may need to modify our GitHub action tests evaluations to prepare a SQL server for testing purposes.
- We may also need to add `clear()` methods to the DAOs in order to test them properly.

To address these concerns I did the following:
- Added to James Submission DAO test files cases for all Submission DAO methods on the interface
- Made examples of the testdb.env file with the values that would be required for a developer to run the DAO tests locally
- Modified the workflow to initialize a database
- Created a DAOTestUtils file that sets up the database. In the future, we can make a generalized function to clear the database. This way, clear functionality stays out of the production code.

## Testing

- [x] Added/updated unit tests
- [x] Tested edge cases
- [x] Manual testing (if needed) - I ran a few submissions through my local autograder to make sure that connection pooling didn't affect student submissions (it didn't)

## Dependencies

This work was begun to test the changes made in #542. However, the tests should pass with/without those changes.

To achieve the performance required for these tests, I added a library [HikariCP](https://github.com/brettwooldridge/HikariCP) which pools MySQL connections to the database. This greatly improved test time, and is likely to improve production performance as well.

## Future Work

(Still applicable) If this goes well, we should generalize the tests for other DAO classes. I'm thinking:
* ConfigurationDAO
* QueueDAO
* RepoUpdateDAO
* RubricConfigDAO
* UserDAO

We should also make sure the tests are running against the MemoryDAOs, SqlDAOs, and more generally, _all DAOS_ available for a particular object.

Still needs to be fixed. Also, the clear method in the tests should later be changed to use a generalized clear method that should be placed int the DAOTestUtils.

In addition to these, there may be some tweaking we could do on the connection pool to optimize it. Currently it is running with default setting, but seeing how our application is pretty small, we may be able to scale down some of the settings. I'm not sure what the architecture of the EC2 we have running though, so I'm going to leave this as the default values for now.

## Additional Notes

~~Please help me! I need some support getting these tests up and running.~~

Thank you all for your feedback through the process of getting this running! Special thanks to @jerodw for the suggestion of using a connection pool!

### CODE REVIEW

**Review by Maillman — COMMENTED** [review]
I haven't run these tests myself, but I have provided some more specific feedback on what I have seen so far.

## Running the tests as a Developer
> When I tried to get this to work with docker though, it didn't detect any variables. I did edit the docker compose (not committed or pushed, so I guess I could have done something wrong) and yet the docker image didn't seem to pick up the environment variables. Am I on the right track? Or is there a better way that anyone knows to test our SQL DAOs? (EDIT: the docker image edit didn't come to be, but if anyone knows a way to get this to work, it might be easier than the current implementation)

I unfortunately don't know of a better or easier way to test the DAOs at the moment. But if there is any, I'm sure it could be figured out with enough trial and research!

## Test Dependencies
> These tests are in no way isolated from each other. I guarantee you if something is wrong with insertSubmission, the whole file will come crashing down. I'm not sure if there's a better way to test the DAO though. Is this a problem?

Hmm... I see what you mean. Every test needs a certain amount of data to be in the table in order to test the DAO's other methods properly, so I don't know if there is a way around this. I don't believe this should be a problem. As far as I'm concerned, I want to make sure insertSubmission works, before making sure everything else works.

## Random Chance of failure in getAllLatestSubmissions
I don't think we should write extra code in the DAO just so a test can be happy, when it won't be needed in practice. So I would like it if there was a way to make a more deterministic time randomizer or something like that.

## Best Submission Tiebreaker
Do we need a tiebreaker? If we do need a tiebreaker, I'd say the tiebreaker should be the submission submitted the latest.

## Performance
> What do we think is reasonable performance for DAO tests?

I am unsure... Honestly, I'm kind of curious enough to do some research here to see what a reasonable performance for these kinds of tests would be. But I think what we have right now should be sufficient enough for our needs, and if we need to improve the performance later, we can!

## Overall
This is looking great so far, Makenna! Keep up the work!

File: src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.io.FileReader;
+import java.util.Properties;
+import java.util.Scanner;
+
+public class DaoTestUtils {
+
+    /**
+     * In order to successfully run these tests on your machine, you're going to need either some environment
+     * variables, or a file containing them named 'testdb.env'. You'll need the following values:
```
[the remark this answers is not in the document]
Comment (mewilker) [reply src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java:14]: I just put it at the root of the project, but maybe for build purposes it should go in the resources folder.

[the remark this answers is not in the document]
Comment (19mdavenport) [reply src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java:14]: @mewilker If you think it would be helpful, you could create an example file with the keys but without the values in it (or fake values)

File: src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     * <ul>
+     *     <li>DB_HOST</li>
+     *     <li>DB_PORT</li>
+     *     <li>DB_NAME</li>
+     *     <li>DB_USER</li>
+     *     <li>DB_PASS</li>
+     * </ul>
+     * It's highly recommended that you use a different DB name than the one you use for development,
+     * as the tests will drop tables when they need to
```
[the remark this answers is not in the document]
Comment (mewilker) [reply src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java]: I did mean to say necessary, thanks

File: src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (19mdavenport) [reply src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java:14]: @mewilker I'm not completely sure, maybe, but it may also prevent other changes you would want to make. One thing I know that works is naming it something different (like `testdb.example.env`) and telling people to copy the file and name it the specific name before putting in the info

[the remark this answers is not in the document]
Comment (mewilker) [reply src/test/java/edu/byu/cs/dataAccess/DaoTestUtils.java:14]: Looks like Intellij's Github plugin is ignoring any changes I make to the testdb.env file. Seems like this is a valid solution!

**Comment by mewilker** [comment]
After reviewing the code on a high level with Dr. Wilkerson, we discussed that the performance could be greatly improved by reusing connections from a connection pool. This is not a concept we go over in 240, so it makes sense that the code currently grabs a new connection every time, rather than reusing connections. He also suggested using indexes in the table, which may or may not already be happening.

**Comment by mewilker** [comment]
I think the Vue test failed to some networking error, as none of my code edited anything related to the frontend...

**Comment by Maillman** [comment]
I wasn't able to run the tests on my machine because I didn't have a database named 'autotest'.

SubmissionDAOTestNoDatabase

It throws the error in setupConnectionPool():

Full Stack Trace:

```
com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Unknown database 'autotest'

	at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:605)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:585)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:101)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:80)
	at edu.byu.cs.dataAccess.sql.SqlDb.setupConnectionPool(SqlDb.java:153)
	at edu.byu.cs.dataAccess.sql.SqlDb.setUpDb(SqlDb.java:41)
	at edu.byu.cs.dataAccess.DaoTestUtils.prepareSQLDatabase(DaoTestUtils.java:55)
	at edu.byu.cs.dataAccess.SubmissionDaoTest.prepareDatabase(SubmissionDaoTest.java:51)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.sql.SQLSyntaxErrorException: Unknown database 'autotest'
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:121)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.ConnectionImpl.setDatabase(ConnectionImpl.java:2080)
	at com.mysql.cj.jdbc.ConnectionImpl.setCatalog(ConnectionImpl.java:2028)
	at com.zaxxer.hikari.pool.PoolBase.setupConnection(PoolBase.java:437)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:373)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:205)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:483)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:571)
	... 8 more
```

**Comment by TheDavSmasher** [comment]
@mewilker I'm not entirely sure but I believe that if you remove the `config.setCatalog(DB_NAME)` from setupConnectionPool(), line 145, it should fix the problem?

Also, unrelated, but why is every single statement using a duplicate try block structure when only the actual SQL string changes?

**Comment by TheDavSmasher** [comment]
Also, having the .env file be in a resources folder, similar to how we tell students to do theirs for their chess projects, would be nice instead of having the file just exist floating around, per se.

**Comment by mewilker** [comment]
@Maillman thanks for testing that for me. I should have fixed it now, will you try again and make sure it works? @TheDavSmasher the testdb.env is now sought after in the test resources folder. The duplicate try block code should be in a PR that looks into #544, but I feel that tests would be better to write before that to help ensure that code does not break.

**Review by Maillman — APPROVED** [review]
Everything looks good to me at this point! All the tests pass on my machine!
SubmissionDAOTestsPass

**Comment by TheDavSmasher** [comment]
@mewilker I'm running into this problem, I'm pretty sure it must have to do with how it's trying to connect to MySQL but I'm unsure what to do to fix this?
image
Lower down it says connection refused

**Review by TheDavSmasher — APPROVED** [review]
Also worked for me now, so it looks good to me!

Although I'm personally not a fan of the TODOs/FIXMEs in the file given the unsolved topics to discuss, but should those be a separate issue? Like, merge these first, then fix those?

I'll let you decide

**Comment by frozenfrank** [comment]
@mewilker Great work! Thanks for carrying this across the finish line 🏁
